### PR TITLE
 coprocessor, dag: implement `in` function.

### PR DESCRIPTION
**Pull request by winoros**
Support pushing  `in` function down to tikv.
PTAL @AndreMouche

### CODE REVIEW

File: src/coprocessor/dag/expr/compare.rs [hunk 1]
```diff
@@ -245,6 +285,37 @@ where
...
+    }
+    let arg0 = arg.unwrap();
+    let mut has_null = false;
+    for exp in others {
+        let arg1 = f(exp)?;
+        if arg1.is_none() {
+            has_null = true;
+            continue;
+        }
+        let cmp_result = get_order(arg0.clone(), arg1.unwrap())?;
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/compare.rs]: Could we use `get_order(&arg0,&arg1)` instead of `clone`?

File: src/coprocessor/dag/expr/compare.rs [hunk 2]
```diff
@@ -429,6 +500,137 @@ mod test {
...
+                Datum::I64(0),
+            ),
+            (
+                ScalarFuncSig::InInt,
+                vec![Datum::I64(1), Datum::I64(2), Datum::I64(1)],
+                Datum::I64(1),
+            ),
+            (
+                ScalarFuncSig::InInt,
+                vec![Datum::I64(1), Datum::I64(2), Datum::Null],
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/compare.rs]: what about case `vec![Datum::I64(1), Datum::I64(2), Datum::Null, Datum::I64(1)]`?

File: src/coprocessor/dag/expr/compare.rs [hunk 3]
```diff
@@ -245,6 +285,37 @@ where
...
 }
 
+fn do_in<'a, T, E, F>(expr: &'a FnCall, f: F, get_order: E) -> Result<Option<i64>>
+where
+    T: Clone,
+    F: Fn(&'a Expression) -> Result<Option<T>>,
+    E: Fn(T, T) -> Result<Ordering>,
+{
+    let (first, others) = expr.children.split_first().unwrap();
+    let arg = f(first)?;
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: Use `try_opt` instead.

File: src/coprocessor/dag/expr/compare.rs (hunk 2 above)
Comment (winoros) [reply src/coprocessor/dag/expr/compare.rs]: added

File: src/coprocessor/dag/expr/compare.rs [hunk 4]
```diff
@@ -245,6 +290,33 @@ where
     Ok(None)
 }
 
+fn do_in<'a, T, E, F>(expr: &'a FnCall, f: F, get_order: E) -> Result<Option<i64>>
+where
+    T: Clone,
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: I don't think it's necessary anymore.

File: src/coprocessor/dag/expr/compare.rs [hunk 5]
```diff
@@ -245,6 +290,33 @@ where
...
+    F: Fn(&'a Expression) -> Result<Option<T>>,
+    E: Fn(&T, &T) -> Result<Ordering>,
+{
+    let (first, others) = expr.children.split_first().unwrap();
+    let arg = try_opt!(f(first));
+    let mut has_null = false;
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: How about let it be `Ok(Some(0))` by default?

**Comment by winoros** [comment]
PTAL @AndreMouche @BusyJay

File: src/coprocessor/dag/expr/fncall.rs [hunk 6]
```diff
@@ -210,7 +210,14 @@ impl FnCall {
...
             ScalarFuncSig::JsonRemoveSig |
-            ScalarFuncSig::JsonMergeSig => (2, usize::MAX),
+            ScalarFuncSig::JsonMergeSig |
+            ScalarFuncSig::InInt |
+            ScalarFuncSig::InReal |
+            ScalarFuncSig::InString |
+            ScalarFuncSig::InDecimal |
+            ScalarFuncSig::InTime |
+            ScalarFuncSig::InDuration |
+            ScalarFuncSig::InJson => (2, usize::MAX),
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/fncall.rs]: It seems that `select NULL in (select * from test.test)s` will get 0 if `test.test` is empty.

Comment (winoros) [reply src/coprocessor/dag/expr/fncall.rs]: This case won't affect tikv's part.

File: src/coprocessor/dag/expr/compare.rs [hunk 7]
```diff
@@ -429,6 +496,142 @@ mod test {
...
+        let dur1 = Duration::parse(b"01:00:00", 0).unwrap();
+        let dur2 = Duration::parse(b"02:00:00", 0).unwrap();
+        let json1 = Json::I64(11);
+        let json2 = Json::I64(12);
+        let s1 = "你好".as_bytes().to_owned();
+        let s2 = "你好啊".as_bytes().to_owned();
+        let t1 = Time::parse_utc_datetime("2012-12-12 12:00:39", 0).unwrap();
+        let t2 = Time::parse_utc_datetime("2012-12-12 13:00:39", 0).unwrap();
+        let cases = vec![
+            (ScalarFuncSig::CoalesceInt, vec![Datum::Null], Datum::Null),
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: What's the purpose of this case?

Comment (winoros) [reply src/coprocessor/dag/expr/compare.rs]: This is a mistake.

**Comment by winoros** [comment]
PTAL @AndreMouche @BusyJay

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**